internal/validation/validation.go

```diff
@@ -0,0 +1,290 @@
+	if tuple.IsValidObject(user) && !tuple.IsTypedWildcard(user) {
+		return nil
+	}
+
+	// tupleset relation involving a wildcard (covers the '*' and 'type:*' cases)
+	if tuple.IsWildcard(user) {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("unexpected wildcard relationship with tupleset relation '%s#%s'", objectType, relation),
+			TupleKey: tk,
+		}
+	}
+
+	// tupleset relation involving a userset (e.g. object#relation) or a user_id (e.g. not a valid object)
+	if !tuple.IsValidObject(user) {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("unexpected user '%s' with tupleset relation '%s#%s'", user, objectType, relation),
+			TupleKey: tk,
+		}
+	}
+
+	return nil
+}
+
+// validateTypeRestrictions makes sure the type restrictions are enforced.
+// 1. If the tuple is of the form doc:budget#reader@person:bob, then the type "doc", relation "reader" must allow type "person".
+// 2. If the tuple is of the form doc:budget#reader@group:abc#member, then the type "doc", relation "reader" must allow type "group", relation "member".
+// 3. If the tuple is of the form doc:budget#reader@*, we allow it only if the type "doc" relation "reader" allows at least one type (with no relation)
+func validateTypeRestrictions(typesys *typesystem.TypeSystem, tk *openfgapb.TupleKey) error {
+	objectType := tuple.GetType(tk.GetObject())           // e.g. "doc"
+	userType, userID := tuple.SplitObject(tk.GetUser())   // e.g. (person, bob) or (group, abc#member) or ("", *)
+	_, userRel := tuple.SplitObjectRelation(tk.GetUser()) // e.g. (person:bob, "") or (group:abc, member) or (*, "")
+
+	typeDefinitionForObject, ok := typesys.GetTypeDefinition(objectType)
+	if !ok {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("type '%s' does not exist in the authorization model", objectType),
+			TupleKey: tk,
+		}
+	}
+
+	relationsForObject := typeDefinitionForObject.GetMetadata().GetRelations()
+
+	relationInformation := relationsForObject[tk.Relation]
+
+	// case 1
+	if userRel == "" && userID != "*" {
+		for _, typeInformation := range relationInformation.GetDirectlyRelatedUserTypes() {
+			if typeInformation.GetType() == userType {
+				return nil
+			}
+		}
+	} else if userRel != "" { // case 2
+		for _, typeInformation := range relationInformation.GetDirectlyRelatedUserTypes() {
+			if typeInformation.GetType() == userType && typeInformation.GetRelation() == userRel {
+				return nil
+			}
+		}
+	} else if userID == "*" { // case 3
+		for _, typeInformation := range relationInformation.GetDirectlyRelatedUserTypes() {
+			if typeInformation.GetType() != "" && typeInformation.GetRelation() == "" {
+				return nil
+			}
+		}
+	}
+
+	return &tuple.InvalidTupleError{Reason: fmt.Sprintf("User '%s' is not allowed to have relation %s with %s", tk.User, tk.Relation, tk.Object), TupleKey: tk}
+}
+
+// NoopFilterFunc returns a filter function that does not filter out any tuple provided
+// to it.
+func NoopFilterFunc() storage.TupleKeyFilterFunc {
```

Comment:
It's not. I just introduced it because of standard practice of having a Noop implementation of things. I'll remove it though and we can introduce something like it when/if we need it.

Reply:
Removed 👍